forc-pkg/src/source/ipfs.rs

```diff
@@ -0,0 +1,242 @@
+use crate::{
+    manifest::{self, PackageManifestFile},
+    source,
+};
+use anyhow::Result;
+use futures::TryStreamExt;
+use ipfs_api::IpfsApi;
+use ipfs_api_backend_hyper as ipfs_api;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::{
+    fmt,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+use tar::Archive;
+use tracing::{info, warn};
+
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct Cid(cid::Cid);
+
+/// A client that can interact with local ipfs daemon.
+pub type IpfsClient = ipfs_api::IpfsClient;
+
+/// Package source at a specific content address.
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize, Deserialize)]
+pub struct Source(pub Cid);
+
+/// A pinned instance of an ipfs source
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize, Deserialize)]
+pub struct Pinned(pub Cid);
+
+impl Pinned {
+    pub const PREFIX: &'static str = "ipfs";
+}
+
+const PUBLIC_GATEWAY: &str = "https://ipfs.io";
```

Comment:
Discussed offline: let's make this public gateway the default, and allow the user to override the URL in the `Forc.toml` with an optional field called `ipfs-node-url`